Tutorial-Learn-how-to-make-a-backend-change-that-has-an-impact-on-server-data.md

```diff
@@ -0,0 +1,677 @@
+
+*In real-world Oppia development, we would also consider how the UI handles this scenario. For instance, the frontend could validate the character limit before calling the backend, providing immediate feedback to users through a warning message near the text box or a snackbar notification. While this tutorial focuses on backend implementation, incorporating frontend validation would enhance the overall user experience.*
+
+*For this tutorial, we will focus on implementing backend validation. Within the backend, we also need to decide where to add this validation—whether in the service layer, controller layer, domain model layer, or storage model layer. If you find yourself unsure about such decisions in practice, don’t hesitate to reach out to team members for guidance.*
+
+*When implementing this in a real scenario, validation would also need to be enforced in the domain layer's `validate()` method to maintain consistency and adhere to Oppia's standards. For an example, see the `validate()` method in the domain layer: [user\_domain.py\#L217](https://github.com/oppia/oppia/blob/cac148abaaa0bba4d96b9df26aa67fd3068b216c/core/domain/user_domain.py#L217). However, since the focus of this tutorial is on the overall process of making a data-affecting change, we won’t cover domain layer validation in detail here.*
+
+*For this tutorial, we’ll implement the validation in the controller layer.*
+
+> [!IMPORTANT]
+> Practice 6: Add a validation check to ensure the length of the updated bio field in the `PreferencesHandler` class is within the allowed limit before making a call to the service layer to update it in the datastore. 
+>
+> Hint: Explore how validations are implemented in other parts of the codebase to understand the practices followed at Oppia. Pay special attention to how validation methods are structured and where they are called.
+
+To restrict the bio length to 200 characters, we’ll add validation to the `PreferencesHandler`.
+
+In `feconf.py`, add:
+
+```python
+MAX_BIO_LENGTH_IN_CHARS = 200
+```
+
+Update the `put` Method in the Controller by adding a check to enforce the character limit before saving: 
+
+```python
+elif update_type == 'user_bio':
+               self.__validate_data_type(update_type, str, data)
+               if len(data) > feconf.MAX_BIO_LENGTH_IN_CHARS:
+                   raise self.InvalidInputException(
+                       'User bio exceeds maximum character limit: %s'
+                       % feconf.MAX_BIO_LENGTH_IN_CHARS)
+               user_settings.user_bio = data
+```
+
+***Note**: Normally, we would use schema validation to enforce this (e.g., using a regex as demonstrated [here](https://github.com/oppia/oppia/blob/cac148abaaa0bba4d96b9df26aa67fd3068b216c/core/schema_utils.py#L700)). However, the `preferences` handler hasn’t been set up for schema validation yet. Adding schema validation would require defining validations for the entire handler, which is beyond the scope of this tutorial.*
```

Comment:
Would it be better to link to https://github.com/oppia/oppia/wiki/Schemas#how-to-write-validation-schemas-for-handlers instead (updating the surrounding documentation as needed)?